File: problem_2/fibonacci.go
package main

import (
  "fmt"
  "math"
  "time"
)

func main() {
  // Finding the sum of even Fibonacci numbers up to 4,000,000
  t0 := time.Now()
  var N float64
  N = 4000000
  fmt.Println(fibonacci_sum(N))
  t1 := time.Now()
  fmt.Println("Finished in", t0.Sub(t1))
}

// FAST FIBONACCI
func fast_fibonacci(n float64) (float64) {
  var fib float64 
  if n == 0 {
    fib = 0
  } else if n == 1 {
    fib = 1
  // One more case below, intended to mitigate stack overflow
  } else if n == 2 {
    fib = 1
  } else {
    if math.Mod(n, 2) == 0 {
      n = n/2
      fib = fast_fibonacci(n)*(2*fast_fibonacci(n + 1) - fast_fibonacci(n))
    } else {
      n = (n - 1)/2
      fib = math.Pow(fast_fibonacci(n + 1), 2) + math.Pow(fast_fibonacci(n), 2)
    }
  }
  return fib
}

func fibonacci_sum(N float64) (float64) {
  var sum, i float64
  for i = 3; i < N; {
    if fast_fibonacci(i) >= 4000000 {
      break
    }
    sum = sum + fast_fibonacci(i)
    i = i + 3
  }
  return sum
}

File: problem_2/Fibonacci.java
import java.lang.*;

public class Fibonacci {
    public static void main(String[] args) {
        int N;
        N = 4000000;
        Fibonacci fibonacci = new Fibonacci();
        System.out.println(fibonacci.fibonacci_sum(N));
    }

    // FAST FIBONACCI
    public double fast_fibonacci(double n) {
        double fib;
        if (n == 0) {
            fib = 0;
        } else if (n == 1 || n == 2) {
            fib = 1;
        } else {
            if (n % 2 == 0) {
              n = n/2;
              fib = fast_fibonacci(n)*(2*fast_fibonacci(n + 1) - fast_fibonacci(n));
            } else {
                n = (n - 1)/2;
                fib = Math.pow(fast_fibonacci(n + 1), 2) + Math.pow(fast_fibonacci(n), 2);
            }
        }
        return fib;
    }

    public double fibonacci_sum(float N) {
        double sum = 0;
        for (int i = 3; i < N; i += 3) {
            if (fast_fibonacci(i) >= 4000000) {
              break;
            }
            sum = sum + fast_fibonacci(i);
        }
        return sum;
    }
}
